### DISCUSSION TITLE
atoti-jupyterlab3 extension not seen by jupyterlab when installed via poetry

### DISCUSSION THREAD

**Discussion by printhellohetal** [atoti/atoti#788] [General]
TL;DR: It seems like there is an issue for those who use poetry with the after-installation copy for atoti-jupyterlab3 to the appropriate jupyter/labextensions folder. Maybe?

In my escapades upgrading to Atoti v0.8.0 (or Atoti v0.8.1) I notice the atoti-jupyterlab3 extension is claimed to be installed by poetry, but not visible in my JupyterLab setting. In doing some digging, I noticed when using poetry, it doesn’t seem to copy the relevant folder to my <POETRY_CACHE_DIR>/virtualenvs/<environment name>/share/jupyter/labextensions

So I ran two experiments, both new environments.

**Experiment 1: poetry-alone**
New poetry environment, everything installed via poetry.
**Result:** no Atoti jupyterlab extension (see below screenshot), no <POETRY_CACHE_DIR>/virtualenvs/<environment name>/share/jupyter/labextensions/atoti-jupyterlab3

if you want to try for yourself, use something like:
```
mkdir poetry-alone && cd poetry-alone
poetry init
```
run through all the prompts
```
poetry install
poetry run jupyterlab
```

I got this, no lovely "**A**" icon on the LHS.
<img width="720" alt="poetry-alone" src="https://github.com/atoti/atoti/assets/83311196/eb1c42eb-9af0-434a-86f2-645cbb366aef">

**Experiment 2: poetry-and-pip**
New poetry environment, everything but atoti-jupyterlab3 installed via poetry, atoti-jupyterlab3 installed in that environment via pip
**Result:** Atoti jupyterlab extension enabled (see below screenshot), and <POETRY_CACHE_DIR>/virtualenvs/<environment name>/share/jupyter/labextensions/atoti-jupyterlab3 exists

if you want to try for yourself, use something like:
```
mkdir poetry-and-pip && cd poetry-and-pip
poetry init
```
run through all the prompts, don’t add atoti-jupyterlab3
```
poetry install
poetry run pip install atoti-jupyterlab3
poetry run jupyterlab
```

I got this, with the lovely "**A**" on the LHS
<img width="720" alt="poetry-and-pip" src="https://github.com/atoti/atoti/assets/83311196/38194a6c-a91b-4e99-8102-6a6e2823638f">

And here are both <POETRY_CACHE_DIR>/virtualenvs/<environment name>/share/jupyter/labextensions/ folders
<img width="720" alt="folder-comparison" src="https://github.com/atoti/atoti/assets/83311196/f5ec88fe-ccdd-4ed6-b243-c94a4f4959a9">

Falling down a rabbit hole to see if it is a poetry related issue or an atoti related issue, it seems like there is a known issue with poetry not doing the same thing as pip wrt JupyterLab extensions? At least, that is what is suggested by the following two git issues:
https://github.com/jupyterlab/jupyterlab-git/issues/452
https://github.com/python-poetry/poetry/issues/1615

Which is curious to me, because I don't recall using pip to install the jupyterlab plugin (at least, I don’t think I did-I feel like I would have written out a very long saga about it then if I had to do all of this…)
Also, this specific comment suggests some labextensions do successfully install via poetry… so, headscratcher there:
https://github.com/jupyterlab/jupyterlab-git/issues/452#issuecomment-557213265

And here is more digital ink spilled regarding after-installation copy not occurring automagically
https://github.com/python-poetry/poetry/issues/4013

So… what gives?
And if the answer is this works for everyone but me (and a prospective that uses poetry that is also struggling with getting the jupyterlab extension to show), cool.
But if not, does this warrant a statement in the installation guide?
Or a guide on the Community side?
Or both?
Or some other change?

**Comment by HuifangYeo** [comment]
I had this issue since 0.7.4 and raised [internal ticket](https://activeviam.atlassian.net/jira/software/c/projects/PYTHON/issues/PYTHON-64?filter=reportedbyme&jql=project%20%3D%20%22PYTHON%22%20AND%20reporter%20IN%20%28currentUser%28%29%29%20AND%20text%20~%20%22jupyterlab%22%20ORDER%20BY%20created%20DESC). 

Check out this issue as well: https://github.com/python-poetry/poetry/issues/1615#issue-526503990

**Comment by tibdex** [comment]
Thanks for the investigation Hetal.

Your screenshot shows that the `jupyterlab_pygments` extension's data files are in the right place so it proves that Poetry manages to install _some_ JupyterLab extensions correctly. After some debugging I found a limitation in a dependency of Poetry: https://github.com/pypa/installer/issues/193.

Atoti 0.8.3 will work around that limitation to make its JupyterLab extension installable with Poetry.
